File: FMELibrary.js/CompetitionParser.ts
import { promises as fs } from 'fs';
import { Competition } from './Competition';

/**
 * Parses and manages competition data from binary files.
 */
export class CompetitionParser {
    /** File path */
    filePath: string = '';

    /** File header, 8 bytes */
    header: Buffer = Buffer.alloc(0);

    /** Item count */
    count: number = 0;

    /** List of all items */
    items: Competition[] = [];

    /**
     * Initializes a new instance of the CompetitionParser class.
     * @param path - The file path of the source data
     * @param buffer - The buffer containing the competition data
     */
    private constructor(path: string, buffer: Buffer) {
        this.filePath = path;
        this.header = buffer.slice(0, 8);
     this.count = buffer.readInt16LE(8);
        this.items = [];
    }

    /**
 * Asynchronously loads competition data from the specified file path.
  * @param path - The file path to load competition data from
     * @returns A promise that resolves to the loaded CompetitionParser instance
  */
    static async load(path: string): Promise<CompetitionParser> {
      const buffer = await fs.readFile(path);
        const parser = new CompetitionParser(path, buffer);

        let offset = 10; // Skip header (8) and count (2)
        
        while (offset < buffer.length) {
      const result = Competition.fromBuffer(buffer, offset);
   parser.items.push(result.competition);
         offset = result.offset;
        }

        return parser;
    }

    /**
     * Converts the competition data to a buffer for serialization.
  * @returns A buffer representing the serialized competition data
     */
    toBuffer(): Buffer {
        const buffers: Buffer[] = [];
      
        buffers.push(this.header);
        
        const countBuffer = Buffer.alloc(2);
        countBuffer.writeInt16LE(this.items.length, 0);
        buffers.push(countBuffer);

  for (const item of this.items) {
            buffers.push(item.toBuffer());
   }

        return Buffer.concat(buffers);
    }

    /**
     * Asynchronously saves data back to file path.
     * @param filepath - Optional file path. If null, saves to the original file path
   * @returns A promise that resolves when the save operation is complete
     */
    async save(filepath?: string): Promise<void> {
     const buffer = this.toBuffer();
     await fs.writeFile(filepath ?? this.filePath, buffer);
    }
}

File: FMMLibrary/CompetitionParser.cs
namespace FMMLibrary
{
    /// <summary>
    /// Parses and manages competition data from binary files.
    /// </summary>
    public class CompetitionParser : IParser<Competition>
    {
        private readonly List<Competition> items;

        /// <summary>
        /// File path
        /// </summary>
        public string FilePath { get; set; }

        /// <summary>
        /// File header, 8 bytes
        /// </summary>
        public byte[] Header { get; set; }

        /// <summary>
        /// Original item count when loading the file.
        /// </summary>
        public short Count { get; set; }

        /// <summary>
        /// List of all items
        /// </summary>
        public IReadOnlyList<Competition> Items => items.AsReadOnly();

        /// <summary>
        /// Initializes a new instance of the <see cref="CompetitionParser"/> class.
        /// </summary>
        /// <param name="path">The file path of the source data.</param>
        /// <param name="reader">The binary reader containing the competition data.</param>
        private CompetitionParser(string path, BinaryReaderEx reader)
        {
            FilePath = path;
            Header = reader.ReadBytes(8);
            Count = reader.ReadInt16();
            items = new List<Competition>(Count);
        }

        /// <summary>
        /// Asynchronously loads competition data from the specified file path.
        /// </summary>
        /// <param name="path">The file path to load competition data from.</param>
        /// <returns>A task that represents the asynchronous operation. The task result contains the loaded <see cref="CompetitionParser"/> instance.</returns>
        public static async Task<CompetitionParser> Load(string path)
        {
            using var fs = File.OpenRead(path);
            using var reader = new BinaryReaderEx(fs);
            var parser = new CompetitionParser(path, reader);

            while (fs.Position < fs.Length)
            {
                var item = new Competition(reader);
                parser.items.Add(item);
            }

            return await Task.FromResult(parser);
        }

        /// <summary>
        /// Adds the specified competition to the collection.
        /// </summary>
        /// <param name="item">The competition to add to the collection.</param>
        public void Add(Competition item)
        {
            items.Add(item);
            Count++;
        }

        /// <summary>
        /// Converts the competition data to a byte array for serialization.
        /// </summary>
        /// <returns>A byte array representing the serialized competition data.</returns>
        public byte[] ToBytes()
        {
            using var stream = new MemoryStream();
            using var writer = new BinaryWriterEx(stream);

            writer.Write(Header);
            writer.Write((short)Items.Count);
            foreach (var item in Items)
            {
                item.Write(writer);
            }

            return stream.ToArray();
        }

        /// <summary>
        /// Save data back to file path
        /// </summary>
        /// <param name="filepath">Optional file path. If null, saves to the original file path.</param>
        /// <returns>A task that represents the asynchronous save operation.</returns>
        public async Task Save(string? filepath = null)
        {
            var bytes = ToBytes();
            await File.WriteAllBytesAsync(filepath ?? FilePath, bytes);
        }
    }
}
